ext/SpeedyWeatherEnzymeExt.jl

```diff
@@ -0,0 +1,115 @@
+    if iseven(n_real)
+        return [1; [2 for i=2:(n_freq-1)]; 1]
+    else 
+        return [1; [2 for i=2:n_freq]]
+    end 
+end 
+
+### Custom rule for _fourier!(f_north, f_north, grid, S)
+function augmented_primal(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, 
+    f_north::Duplicated, f_south::Duplicated, grids::Duplicated{<:AbstractGridArray}, S::Union{Const, MixedDuplicated}) 
+
+    func.val(f_north.val, f_south.val, grids.val, S.val) # forward pass
+
+    # save grids in tape if grids will be overwritten
+    if overwritten(config)[4] # TODO: Not sure this is really necessary because grids won't ever get overwritten by this _fourier!
+        tape = copy(grids.val)
+    else
+        tape = nothing
+    end
+
+    return AugmentedReturn(nothing, nothing, tape) # because the function actually returns nothing
+
+end 
+
+function reverse(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, tape,
+    f_north::Duplicated, f_south::Duplicated, grids::Duplicated{<:AbstractGridArray}, S::Union{Const, MixedDuplicated})
+
+    # adjoint/vjp of FFT has a different scaling, compute it, apply it later to f_north, f_south
+    scale = adjoint_scale(S.val)
+    
+    # retrieve grids value, either from original grids or from tape if grids may have been overwritten.
+    gridsval = overwritten(config)[4] ? tape : grids.val
+
+    # compute the actual vjp
+    dgridval = zero(gridsval)
+    _fourier!(dgridval, f_north.dval ./ scale, f_south.dval ./ scale, S.val) # inverse FFT (w/o normalization)
+    grids.dval .+= dgridval 
+
+    # no derivative wrt the f_north and f_south that were input because they are overwritten
+    make_zero!(f_north.dval) 
+    make_zero!(f_south.dval)
+
+    # the function has no return values, so we also return nothing here
+    return (nothing, nothing, nothing, nothing)
+end
+
+### Custom rule for _fourier!(grid, f_north, f_south, S)
+function augmented_primal(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, 
+    grids::Duplicated{<:AbstractGridArray}, f_north::Duplicated, f_south::Duplicated, S::Union{Const, MixedDuplicated}) 
+
+    func.val(grids.val, f_north.val, f_south.val, S.val) # forward pass
+
+    # TODO: make an overwritten check here? 
+
+    return AugmentedReturn(nothing, nothing, nothing) # because the function actually returns nothing
+
+end 
+
+function reverse(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, tape,
+    grids::Duplicated{<:AbstractGridArray}, f_north::Duplicated, f_south::Duplicated, S::Union{Const, MixedDuplicated})
+
+    # adjoint/vjp of FFT has a different scaling, compute it, apply it later to f_north, f_south
+    scale = adjoint_scale(S.val)
+    
+    # TODO: retrieve from tape here if overwritten? 
+
+    # compute the actual vjp
+    dfnorthval = zero(f_north.val)
+    dfsouthval = zero(f_south.val)
```

Comment:
Because these allocate: Is that because we can't reuse the forward memory for the gradient?

Reply:
I have no idea to be honest. I didn't do it initially because I am a bit careful with mutating the arguments within things the AD might still need. But I just tested to use `f_north.val` as a buffer, and this seems to work. At least the tests pass. But I don't know if this is generally something we can do. I didn't commit this yet to the PR.